sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -3819,8 +3841,29 @@ piEnqueueKernelLaunch(pi_queue Queue, pi_kernel Kernel, pi_uint32 WorkDim,
 
   ZE_CALL(zeKernelSetGroupSize, (Kernel->ZeKernel, WG[0], WG[1], WG[2]));
 
+  pi_platform Plt = Kernel->Program->Context->Devices[0]->Platform;
   // Lock automatically releases when this goes out of scope.
-  std::lock_guard<std::mutex> lock(Queue->PiQueueMutex);
+  std::lock(Queue->PiQueueMutex, Plt->MemAllocsMutex);
+  std::lock_guard<std::mutex> QueueLock(Queue->PiQueueMutex, std::adopt_lock);
+  std::lock_guard<std::mutex> MemAllocsLock(Plt->MemAllocsMutex,
+                                            std::adopt_lock);
+
+  // We are going to submit a kernel for execution. If indirect access flag is
+  // set for a kernel then we need to make a snapshot of existing memory
```

Comment:
> If indirect access flag is set for a kernel

You aren't checking this. I suggest create an utility function that will for now just return true as we always set this flag, but which can be extended when/if we are provided an interface to query if the indirect access was actually used or backend compiler could prove it wasn't needed and dropped it.

Reply:
Thx, added function.